crates/socket/src/multiplexing.rs

```diff
@@ -131,68 +132,55 @@ where
 
 /// Implement async socket where response are send back async manner
 /// they are queued using channel
+#[pin_project]
 pub struct AsyncResponse<R> {
+    #[pin]
     receiver: Receiver<BytesMut>,
     header: RequestHeader,
     correlation_id: i32,
     data: PhantomData<R>,
 }
 
-impl<R> AsyncResponse<R>
-where
-    R: Request,
-{
-    pub async fn next(&mut self) -> Result<R::Response, FlvSocketError> {
-        debug!(
-            "waiting for async response: {} correlation: {}",
-            R::API_KEY,
-            self.correlation_id
-        );
+impl<R: Request> Stream for AsyncResponse<R> {
+    type Item = Result<R::Response, FlvSocketError>;
 
-        if let Some(res_bytes) = self.receiver.next().await {
-            let response =
-                R::Response::decode_from(&mut Cursor::new(&res_bytes), self.header.api_version())?;
-            trace!("receive response: {:#?}", &response);
-            Ok(response)
-        } else {
-            error!("no more response. server has terminated connection");
-            Err(IoError::new(ErrorKind::UnexpectedEof, "server has terminated connection").into())
-        }
-    }
+    #[instrument(
+        skip(self, cx),
+        fields(
+            api_key = R::API_KEY,
+            correlation_id = self.correlation_id,
+        )
+    )]
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        let next = match this.receiver.poll_next(cx) {
+            Poll::Pending => {
+                debug!("Waiting for async response");
+                return Poll::Pending;
+            }
+            Poll::Ready(next) => next,
+        };
 
-    pub async fn next_timeout(
```

Comment:
maybe this is not good api.  we should just add some timing attribute

Reply:
I think that we can integrate timeouts using some kind of composition with the stream trait or future trait. I figured we didn't need to roll our own for a specific use-case now that we're hooking into the rest of the futures ecosystem